clients/src/main/java/org/apache/kafka/common/AutoCreateTopicsPolicy.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+import java.util.stream.Collectors;
+
+public enum AutoCreateTopicsPolicy {
```

Comment:
Do you think there is much benefit to the `disallowed` option? We'll see fewer and fewer clusters on versions older than 0.11 going forward. Backwards compatibility is only supported for brokers since 0.10.0 anyway, so it is only the 0.10.x brokers that make this tricky. Perhaps it would be sufficient to document the behavior and maybe log a warning if connecting to one of these brokers?

Reply:
This is an interesting question. Realistically, when do we expect to change the default? Could we simply bump the minimum supported broker version then? 0.10.2 is 2 years old at this point. If we change the default in 2.3 or 2.4, it might be reasonable to just drop support for 0.10.x brokers.

Reply:
My thought was to change the default to "disallowed" in 3.0 with the caveat that this would only be enforced for 0.11+ brokers. Dropping support for 0.10.x might be possible, but it might be overkill if this is the only compatibility problem.

Reply:
My thought was we could safely change the default to "disallowed_if_supported_by_brokers" until the client continues to support 0.10.x brokers. When we drop 0.10.x support, both "disallowed" and "disallowed_if_supported_by_brokers" would essentially have the same behavior.

I think the main question is what we want to do now, while we support 0.10.x. Are we okay with silently allowing auto topic creation for 0.10.x brokers or do we want a mechanism that throws an exception instead? The current patch implements both of these and leaves the choice to the user.

Reply:
fwiw, if we had to choose, I think "disallowed" is better to keep than "disallowed_if_supported_by_brokers", more so if we assume there will be very few 0.10.x users. This configuration results in deterministic behavior, we either allow topic creation or not.

We could switch the default to "disallowed" when needed and document the incompatible change for 0.10.x users, who would have to manually override this config.

Reply:
@dhruvilshah3 I agree the argument cuts both ways. Long term, if we assume that we will eventually drop support for 0.10 clients, then it would be a shame to have an unneeded config policy left around. Perhaps we can just choose the behavior we like best for "disallow." If we choose the strict behavior, then dropping support for 0.10 will probably be the prerequisite for making it the default, which may be reasonable. I'm not sure we've ever documented how long the clients should continue working with older brokers.